app/controllers/application_controller.rb

```diff
@@ -4,7 +4,7 @@ class ApplicationController < ActionController::Base
   include SSOIdentity
 
   def authenticate_user
-    unless sso_identity
+    unless sso_identity && !session_expired?
       session[:redirect_path] = request.original_fullpath
```

Comment:
mmm Something not right here methinks ... should we be logging out the user from SSO first?

Reply:
This works okay for me. It does crash on first run of the branch (if you are currently logged in, because your session info doesn't include the expiry field), but it seems to behave the way we discussed.